### DISCUSSION TITLE
How to use a private DNS with k3s

### DISCUSSION THREAD

**Discussion by EugenMayer** [k3s-io/k3s#7822] [Q&A] [answered]
Trying to configure k3s to use my private dns server `10.10.0.243` as the forward/upstream default to resolve. I fail to do so, my pods cannot resolve the private domains, only public ones.

Using ubuntu 22.04 with systemd based DNS.

The only way i could get it running is defining a `coredns-custom` custom-map like this

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns-custom
  namespace: kube-system
data:
  custom.server: |
     myprivatedomain.tld:53 {
      forward . 10.10.0.243
    }
```

Is there a better way to deal with this issue?

**Comment by brandond** [comment]
> Since k3s does not use the OS DNS server to forward DNS requests

It does, unless the host's resolv.conf includes an invalid upstream, in which case it uses 8.8.8.8. If you want to point k3s at an alternative upstream, the easiest way is to create a custom resolv.conf, and point k3s at it with the `--resolv-conf` flag.

**Comment by EugenMayer** [comment]
@brandond you have answered this way several times, but there is more to it:

- we have a valid upstream - i mean it is 127.0.0.53 since it is ubuntu / systemd based DNS, but the configuration is perfectly valid. Still k3s picks 8.8.8.8 (but the way, this does work in rke2). I assume k3s determines that `127.0.0.53` is problematic form within the coredns container and skips it. It should instead just use the node ip - i mean this is normal for all new systemd / network-manager setups and will not go away.
- using `--resolv-conf` will only help until the restart tight, AFAICU the systemd unit will not include that param and one has to add the param to the unit file on every upgrade right?

Right now, the only way i could "fix" it is using

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns-custom # this is the name of the configmap you can overwrite with your changes
  namespace: kube-system
data:
    custom.server: | 
    prviatedomain.tld:53 {
       forward . <provate DNS IP>
    }
```

which i would need to do for every private-domain i need to follow.

Happy to understand your POV better

**Reply by brandond** [reply]
> we have a valid upstream - i mean it is 127.0.0.53 since it is ubuntu / systemd based DNS

That is not a valid upstream; 127.x.x.x in a pod reaches that pod, not the host.

> It should instead just use the node ip

Its not any safer of an assumption to expect the node IP to work as a DNS server, in the case of systemd-resolved which only listens on that loopback address.

> using --resolv-conf will only help until the restart tight, AFAICU the systemd unit will not include that param and one has to add the param to the unit file on every upgrade right?

Only if you are upgrading by re-running the install script, and don't have some way to ensure that you consistently pass the same arguments every time. You could use the config file, which isn't managed by the install script.

> Right now, the only way i could "fix" it is using

Or you could just create a custom resolv.conf that points at the real upstream that is hosting your internal domains, which will presumably also act as a recursive resolver for public zones.

**Comment by EugenMayer** [comment]
Thank you for elaborating and thank you for helping!

Considering that ubuntu LTS is one of the official k8s distros and k3s not supporting the default cloud-init based installation is something that i really consider is a issue. I understood why you cannot use the default resolv.conf it and i also understand that most probably less people use a private DNS (thus e.g. like throwing an error on start is not desired).

To reduce the amount of issues for that questions (there are a couple already), tell people why the syst. Not just a generic "if it is invalid" - this would help reducing the amount of issues opened in k3s (probably)

i really would wish there would be a more elegant way to work with that issue. The only one i can think of technically is either

- loopback device alias like `172.31.31.254` to the loopback and address that as the DNS resolver
- habe a k3s config way of setting the resolv.conf or just the forward dns (probably already possible?)

**Reply by brandond** [reply]
> k3s not supporting the default cloud-init based installation is something that i really consider is a issue.

I'm not sure what any of this has to do with cloud-init?

when systemd-resolved is in use, there is normally an unaltered resolv.conf at `/run/systemd/resolve/resolv.conf` that points to the actual upstream resolver. K3s will check both `/etc/resolv.conf`, and `/run/systemd/resolve/resolv.conf`, and use whichever one contains a viable upstream resolver. Only when neither file is viable does k3s use 8.8.8.8.

> habe a k3s config way of setting the resolv.conf

There is such a way, I mentioned it above: use the --resolv-conf flag.

**Reply by dragonfleas** [reply]
> when systemd-resolved is in use, there is normally an unaltered resolv.conf at `/run/systemd/resolve/resolv.conf` that points to the actual upstream resolver. K3s will check both `/etc/resolv.conf`, and `/run/systemd/resolve/resolv.conf`, and use whichever one contains a viable upstream resolver. Only when neither file is viable does k3s use 8.8.8.8.

I know this is an old thread but could this be added to the docs anywhere? Maybe inside of https://docs.k3s.io/networking/basic-network-options? It would be really useful because it took me ages to dig for what K3S looks for in particular. This is a hard requirement that is seemingly undocumented, besides in obscure github PRs and discussions.

Also where in the CoreFile does it specify that it searches `/run/systemd/resolve/resolv.conf`? I am really confused on where that is coming from. If there's explicitly _only_ `/etc/resolv.conf` in the CoreFile, is CoreDNS the service taking care of the fallback systemd run conf?

**Reply by dragonfleas** [reply]
To expand on this, this is taken from the [loop plugin readme for CoreDNS](https://github.com/coredns/coredns/blob/master/plugin/loop/README.md)
> There are many ways to work around this issue, some are listed here:
> 
> Add the following to your kubelet config yaml: resolvConf: <path-to-your-real-resolv-conf-file> (or via command line flag --resolv-conf deprecated in 1.10). Your "real" resolv.conf is the one that contains the actual IPs of your upstream servers, and no local/loopback address. This flag tells kubelet to pass an alternate resolv.conf to Pods. For systems using systemd-resolved, /run/systemd/resolve/resolv.conf is typically the location of the "real" resolv.conf, although this can be different depending on your distribution.
>
> Disable the local DNS cache on host nodes, and restore /etc/resolv.conf to the original.
> A quick and dirty fix is to edit your Corefile, replacing forward . /etc/resolv.conf with the IP address of your upstream DNS, for example forward . 8.8.8.8. But this only fixes the issue for CoreDNS, kubelet will continue to forward the invalid resolv.conf to all default dnsPolicy Pods, leaving them unable to resolve DNS.

It doesn't state that it will check `/run/systemd/resolve/resolv.conf` so does k3s provide the abstraction for this somewhere in the kubelet definition?

**Comment by EugenMayer** [comment]
This is interesting, this is my `/run/systemd/resolve/resolv.conf` 

```
nameserver 10.10.0.243
nameserver 1.1.1.1
search .
```

It is never used by k3s. Is the `search .` the actual issue then?

> k3s not supporting the default cloud-init based installation is something that i really consider is a issue.

Ubuntu 20.04/22.04 based setups which are e.g. booted using cloud-init (kind of common for setting up k3s/rke2) come with the describe systemd based dns setup. The statement you made about `/run/systemd/resolve/resolv.conf` made this irrelevant, since this would support those setups too.

**Reply by brandond** [reply]
You can see the checks K3s performs at:
* https://github.com/k3s-io/k3s/blob/master/pkg/agent/config/config.go#L278-L300
* https://github.com/k3s-io/k3s/blob/master/pkg/agent/config/config.go#L306-L311

It doesn't care about the search domain, just whether or not a nameserver entry in the file points at a address that is not a global unicast address, as documented at https://pkg.go.dev/net#IP.IsGlobalUnicast

You should be able to see in the k3s logs which resolv.conf it is using in the kubelet args, as those are printed out during startup.

**Reply by EugenMayer** [reply]
AFAICS this would mean that my `/run/systemd/resolve/resolv.conf` with 

```
nameserver 10.10.0.243
nameserver 1.1.1.1
search .
```

Should have been used (which it did not). Should i open a bug for this or did i misunderstood you here?

**Reply by brandond** [reply]
Please post the logs showing what the kubelet is actually being configured to use.

There isn't really much potential for bugs here, the code is quite simple in what it looks for. From that point on, the behavior is dictated by the kubelet and coredns.

**Reply by EugenMayer** [reply]
With logs you most probably some CoreDNS config maps? Which one doe you need in particular?

**Reply by brandond** [reply]
no, the logs from k3s starting up. It will contain a line that looks like this - note that the `--resolv-conf` value is visible.

`INFO[0002] Running kubelet --address=0.0.0.0 --allowed-unsafe-sysctls=net.ipv4.ip_forward,net.ipv6.conf.all.forwarding --anonymous-auth=false --authentication-token-webhook=true --authorization-mode=Webhook --cgroup-driver=cgroupfs --client-ca-file=/var/lib/rancher/k3s/agent/client-ca.crt --cloud-provider=external --cluster-dns=10.43.0.10 --cluster-domain=cluster.local --container-runtime-endpoint=unix:///run/k3s/containerd/containerd.sock --containerd=/run/k3s/containerd/containerd.sock --eviction-hard=imagefs.available<5%,nodefs.available<5% --eviction-minimum-reclaim=imagefs.available=10%,nodefs.available=10% --fail-swap-on=false --healthz-bind-address=127.0.0.1 --hostname-override=k3s-server-1 --kubeconfig=/var/lib/rancher/k3s/agent/kubelet.kubeconfig --kubelet-cgroups=/k3s --node-labels= --pod-infra-container-image=rancher/mirrored-pause:3.6 --pod-manifest-path=/var/lib/rancher/k3s/agent/pod-manifests --read-only-port=0 --resolv-conf=/etc/resolv.conf --runtime-cgroups=/k3s --serialize-image-pulls=false --tls-cert-file=/var/lib/rancher/k3s/agent/serving-kubelet.crt --tls-private-key-file=/var/lib/rancher/k3s/agent/serving-kubelet.key`

**Reply by EugenMayer** [reply]
Anytime:

```
time="2023-06-25T22:49:00Z" level=info msg="Running kubelet --address=0.0.0.0 --allowed-unsafe-sysctls=net.ipv4.ip_forward,net.ipv6.conf.all.forwarding --anonymous-auth=false --authentication-token-webhook=true --authorization-mode=Webhook --cgroup-driver=systemd --client-ca-file=/var/lib/rancher/k3s/agent/client-ca.crt --cloud-provider=external --cluster-dns=10.43.0.10 --cluster-domain=cluster.local --container-runtime-endpoint=unix:///run/k3s/containerd/containerd.sock --containerd=/run/k3s/containerd/containerd.sock --eviction-hard=imagefs.available<5%,nodefs.available<5% --eviction-minimum-reclaim=imagefs.available=10%,nodefs.available=10% --fail-swap-on=false --healthz-bind-address=127.0.0.1 --hostname-override=solo-k8s-k3s-prod --kubeconfig=/var/lib/rancher/k3s/agent/kubelet.kubeconfig --node-labels= --pod-infra-container-image=rancher/mirrored-pause:3.6 --pod-manifest-path=/var/lib/rancher/k3s/agent/pod-manifests --read-only-port=0 --resolv-conf=/run/systemd/resolve/resolv.conf --serialize-image-pulls=false --tls-cert-file=/var/lib/rancher/k3s/agent/serving-kubelet.crt --tls-private-key-file=/var/lib/rancher/k3s/agent/serving-kubelet.key"
```

**Reply by brandond** [reply]
OK, so k3s is correctly pointing the kubelet at the expected file: `--resolv-conf=/run/systemd/resolve/resolv.conf`

That indicates that the resolv.conf files generated by the kubelet for pods that use `dnsPolicy: Default` (which is what the coredns pods use) would point at the upstream resolvers listed in that file. coredns in turn is configured to resolve things using that resolv.conf via the `forward . /etc/resolv.conf` directive. If the upstreams listed in that file can resolve your private zones, you shouldn't need to do anything at all.

Note that none of this is K3s specific; its all just normal Kubelet and containerd stuff.

**Reply by EugenMayer** [reply]
Just checked, in fact the CoreDNS pod uses `dnsPolicy: Default`.

The pod that is problematic uses `dnsPolicy: ClusterFirst` - looking at https://github.com/stevehipwell/helm-charts/blob/main/charts/nexus3/templates/deployment.yaml this will be the default since it is not set here. Checking the deployment on the cluster and it is as expected `dnsPolicy: ClusterFirst`

The pod will not be able to resolve private DNS entries unitl i deploy a `coredns-custom` with

```
myprivatdomain.tld:53 {
  forward . 10.10.0.243
}
```

This means: 
 - coreDNS has network access to the expted upstream 10.10.0.243 and would be able to forward, if asked
 - are you expecting the pods to be created with `dnsPolicy: Default` or is this just right for the CoreDNS pod? If it is the former, i assume that can be defined as a default somewhere?

**Reply by brandond** [reply]
Generally you want pods to use ClusterFirst, which means they'll use coredns. Coredns should in turn forward everything to the upstreams specified in `/run/systemd/resolve/resolv.conf`, as I mentioned above. coredns randomly balances queries across all listed upstreams - which means that half of queries from your pods will be forwarded to 10.10.0.243 and will resolve your private domain properly, while half will be sent to 1.1.1.1 and will not. Have you tried just taking the public upstream out of `/run/systemd/resolve/resolv.conf`?

**Reply by EugenMayer** [reply]
Balancing?!? Since when that became a standard for nameservers? https://linux.die.net/man/5/resolv.conf does say as expected, they are tried in order and the first one that resolves, is used. Strictly sequential.

So when did CoreDNS decide to skip that and just load balance it? I mean in my case it is just "when the internal is down, try at least the public one" - it is a fallback and rather hard to change for me since the DNS config is pushed by the OpenStack DHCP / network init process.

Checking https://coredns.io/plugins/forward/ i cannot find that CoreDNS uses such a load-balancing behavior - could you reference it somehow? I understand, if that is how CoreDNS behaves, it has nothing to do with k3s.
But somehow i'am yet "reluctant" to believe that CoreDNS changed such a fundamental, especially when specifically asked to use the resolv.conf and it simply uses the nameservers in a different meaning as they are defined in sysops..

Thanks for all your help!

**Reply by brandond** [reply, marked as answer]
> i cannot find that CoreDNS uses such a load-balancing behavior - could you reference it somehow

It's right there in the page you linked...

> TO… are the destination endpoints to forward to. The TO syntax allows you to specify a protocol, tls://9.9.9.9 or dns:// (or no protocol) for plain DNS. The number of upstreams is limited to 15.
> **Multiple upstreams are randomized** (see `policy`) on first use. When a healthy proxy returns an error during the exchange the next upstream in the list is tried.
>
>policy specifies the policy to use for selecting upstream servers. The default is random.
> random is a policy that implements random upstream selection.

The doc doesn't call this out, but it you point it at a resolv.conf instead of listing the upstreams in the config file itself, it will read the resolv.conf to determine what upstreams to use, just as if you had listed them all explicitly in the config file.

I guess it's not true load-balancing, but simply selecting a random upstream, and then using that one until it gets an error (and nxdomain is not an error), at which point the policy determines how it picks the next one to try. We use similar behavior for the k3s agent load-balancer tunnel to servers, so I have a bad habit of referring to this as load-balancing when it is really more of a failover pool.

The behavior stands though; coredns has a 50/50 chance of picking your private upstream when it start up.

> when did CoreDNS decide to skip that and just load balance it

Always, as long as coredns has been around

**Reply by EugenMayer** [reply]
Sorry! Well i understood that 

```
forward . 8.8.8.8 1.1.1.1
```

is randomized but 

```
forward . /etc/resolve.conf
```
 
is telling it to use the resolve.conf and thus go by it's rules. I did not take the latter as "pick all the nameservers from that file and add them as simple upstreams".

But as i see the `policy` can be adjusted to `policy: sequential`.

IMHO if resolve.conf is used, it should become `policy: sequential` since anything else is just redefining the common expectations - but maybe it is just me sticking to the linux defaults, esp. on sysops. I guess there is no consideration in changing that. Kind of sad that it all came down to this in the end. 

in the end
 - `/etc/resolv.conf` is not used, since it is detected to be a systemd-dns environment
 - instead `/run/systemd/resolve/resolv.conf` is used, which includes what i defined as DNS server, including my private DNS
 - the setting is properly propagated by k3s to CoreDNS
 - CoreDNS uses the `dnsPolicy: default` which does instruct it to use the OS
 - but the CoreDNS `policy` setting in coreDNS defaults to "random" and thus my privateDNS server was not used half of the time

I understand that this is the CoreDNS default, surprising for me or not. Thank you for teaching me that!

Expecting k3s to have an extra-souce to fix that issue if `/etc/resolv.conf` is used is maybe not only too much to ask, it's probably not everyone's expectation either.

So for me defining there are 2 options
 - defining a custom resolv.conf for k3s with only one server and use it with `--resolv-conf`
 - try to override the policy in coreDNS `:53`. AFAICS there is no easy way to add that value only (and persist that even after k3s restarts). Current overrides `coredns-custom` only allow to redefine the entire section AFAIU.

Again, thank you for your patience and energy to share all those details. Very appritiated!

**Reply by brandond** [reply]
The releases that just came out this week allow you to override entries in the default server block (see https://github.com/k3s-io/k3s/pull/7583), you should be able to do something like this:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns-custom
  namespace: kube-system
data:
  forward.override: |
    forward . /etc/resolv.conf {
      policy sequential
    }
```

I am not sure what happens if your internal upstream fails though; I suspect that it will switch over to the public upstream and stay there until it returns an error. You'd probably want to just have it forward to only your private upstream.

**Reply by EugenMayer** [reply]
One last question: Should i rephrase the intial question to something like : "How to use a private DNS with k3s" and change my initial question a bit, then mark the summary i gave above as answer?

The summary would be something like

```
Trying to configure k3s to use my private dns server `10.10.0.243` as the forward/upstream default to resolve. I fail to do so, my pods cannot resolve the private domains, only public ones.

Using ubuntu 22.04 with systemd based DNS.

The only way i could get it running is defining a `coredns-custom` custom-map like this

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns-custom
  namespace: kube-system
data:
  custom.server: |
     myprivatedomain.tld:53 {
      forward . 10.10.0.243
    }
---

Is there a better way to deal with this issue?
```

The reason i would like to do that is to value all the effort you put into answering all the qeuestions, details and teach. I think this thread could help someone else too but most of the issues i have read about this in k3s are not elaborating as much at all like you did here IMHO.

I let you decide of course. You helped me already (by a lot) - all i would like is that this might help others too (if anybody is actually as stubborn as iam)

**Reply by EugenMayer** [reply]
> The releases that just came out this week allow you to override entries in the default server block (see #7583), you should be able to do something like this:
> 
> ```yaml
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: coredns-custom
>   namespace: kube-system
> data:
>   forward.override: |
>     forward . /etc/resolv.conf {
>       policy sequential
>     }
> ```
> 
> I am not sure what happens if your internal upstream fails though; I suspect that it will switch over to the public upstream and stay there until it returns an error. You'd probably want to just have it forward to only your private upstream.

Did that

```
kubectl get cm -n kube-system coredns-custom -o yaml
apiVersion: v1
data:
  forward.override: |
    forward . /etc/resolv.conf {
      policy sequential
    }
immutable: false
kind: ConfigMap
metadata:
  name: coredns-custom
  namespace: kube-system
```

one rather odd thing is, it really takes a while after a server reboot for the DNS to actually work.
So after a server reboot, my pods all are started for already quiet some time, but the DNS resolving to private does only work after about 30-60s. Mayeb something quirky on my system - buy for everyone trying out, you might need to be a little patient.

Again, thank you for this really elegant solution!

**Reply by mreiche** [reply]
Hi Eugen. I'm looking for a solution for private DNS setups.  It only works when I change the `coredns` ConfigMap from:

```
forward . /etc/resolv.conf
```
to
```
forward . <My-Private-DNS-IP>
```

I tried to override this block like you did. But it won't work for me. I already restarted k3s and waited some time. Can you explain the solution to me and how it works?

**Comment by brandond** [comment]
> Should i rephrase the intial question to something like : "How to use a private DNS with k3s" and change my initial question a bit, then mark the summary i gave above as answer?

You're welcome to do whatever you think would help people find it best!

**Reply by EugenMayer** [reply]
thank you, did so!

**Comment by EugenMayer** [comment]
If anybody is in need, i created a little helm chart to fix this issue, see https://github.com/EugenMayer/helm-charts/tree/main/charts/coredns-private-dns-fix

**Reply by thematrixdev** [reply]
Hello there.
I am new to K3S and wasted a whole morning figuring out DNS issue...
I have just found this thread and installed your helm chart.
How to make it work?

```
$ helm list -A
NAME                              	NAMESPACE  	REVISION	UPDATED                                	STATUS  	CHART                        	APP VERSION
coredns-private-dns-fix-1738222319	default    	1       	2025-01-30 07:32:00.144059601 +0000 UTC	deployed	coredns-private-dns-fix-0.1.1	0.1.0
```

`$ kubectl edit configmap coredns -n kube-system`
```
apiVersion: v1
data:
  Corefile: |
    .:53 {
        errors
        health
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        hosts /etc/coredns/NodeHosts {
          ttl 60
          reload 15s
          fallthrough
        }
        prometheus :9153
        forward . 10.28.0.2 # Route 53
        cache 30
        loop
        reload
        loadbalance
        import /etc/coredns/custom/*.override
    }
    import /etc/coredns/custom/*.server
  NodeHosts: |
    10.28.105.0 k3s-server
    10.28.107.64 k3s-agent-1
kind: ConfigMap
metadata:
...
```

```
$ kubectl rollout restart deployment coredns -n kube-system

$ kubectl run -it --rm --restart=Never --image=busybox --namespace=my_ns test-pod -- /bin/sh

# nslookup stg.mariadb.private

;; connection timed out; no servers could be reached

# nslookup stg.mariadb.private 10.28.0.2
Server:		10.28.0.2
Address:	10.28.0.2:53

Non-authoritative answer:
Name:	stg.mariadb.private
Address: 10.28.114.247

Non-authoritative answer:
```

**Reply by EugenMayer** [reply]
Applying the helm chart and restartin the node should usually do it. Beside that, do not use an alpine pod to test. Because it's alpine ;)

**Reply by thematrixdev** [reply]
You are correct.

I have found in your helm chart:
```
forward . /etc/resolv.conf {
  policy sequential
}
```

So it'd be better to use the default configuration `/etc/resolv.conf`.

`/etc/resolv.conf` on K3S server-node host:
```
nameserver 127.0.0.53
options edns0 trust-ad
search .
```

`/etc/systemd/resolved.conf` on K3S server-node host:
```
DNS=10.28.0.2
```

There is no need to `kubectl edit configmap coredns -n kube-system`.

Thanks.

**Comment by zalmane** [comment]
I wanted to add a case that isn't covered by the customer /etc/resolv.conf which is using a local resolver such as dnsmasq. in that case we are forced to put in the private DNS since the /etc/resolv.conf shows 127.0.0.1. 
Is there any solution to this other than manually putting in the DNS servers?

**Reply by mreiche** [reply]
I'm currently asking the same question here: https://github.com/k3s-io/k3s/issues/10951

**Comment by isc30** [comment]
Hey, I can't get it to work as expected, even after adding the workarounds like the ConfigMap and the helm chart. Even after restarting the coredns pod and the nodes. It still uses random nameservers, any ideas?

My `/etc/resolv.conf` looks like this:
```
nameserver 192.168.1.3
nameserver 1.1.1.1
```

and running `nslookup` inside a pod:
```sh
/ # nslookup k3s-stone.dns.my.casa 192.168.1.3
Server:         192.168.1.3
Address:        192.168.1.3:53

Non-authoritative answer:
Name:   k3s-stone.dns.my.casa
Address: 192.168.20.13

/ # nslookup k3s-stone.dns.my.casa 1.1.1.1
Server:         10.43.0.10
Address:        10.43.0.10:53

k3s-stone.dns.my.casa  canonical name = my.casa
Name:   my.casa
Address: 192.168.20.2
```

Randomly shows the correct `192.168.20.13` and the incorrect `192.168.20.2` ones.
The incorrect one is coming from `1.1.1.1`, so its not respecting the order:

```sh
/ # nslookup k3s-stone.dns.my.casa
Server:         10.43.0.10
Address:        10.43.0.10:53

k3s-stone.dns.my.casa  canonical name = my.casa
Name:   my.casa
Address: 192.168.20.2

/ # nslookup k3s-stone.dns.my.casa
Server:         10.43.0.10
Address:        10.43.0.10:53

Non-authoritative answer:
Name:   k3s-stone.dns.my.casa
Address: 192.168.20.13
```

**Reply by isc30** [reply]
I'm using uptime kuma to send ping requests to these hostnames and it shows that coredns is still alternating upstreams:
![image](https://github.com/user-attachments/assets/321cc261-8509-49bc-ac01-0b97dd2649c8)

**Reply by isc30** [reply]
The only way I got it to work is by manually defining all "my domains" to use the private DNS. Does anyone know why the `forward.override` isn't working but `custom.server` works?
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns-custom
  namespace: kube-system
data:
  custom.server: | 
    my.casa:53 {
      forward . /etc/resolv.conf {
        policy sequential
      }
    }
    another.domain:53 {
      forward . /etc/resolv.conf {
        policy sequential
      }
    }
```

**Reply by isc30** [reply]
Solution, this works:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: coredns-custom
  namespace: kube-system
data:
  forward.override: |
    forward . /etc/resolv.conf {
      policy sequential
    }
```

**Comment by d-a-n-k-a** [comment]
one of the options is to edit ConfigMap in the `/var/lib/rancher/k3s/server/manifests/coredns.yaml`

**Reply by brandond** [reply]
no, don't do that. changes will be reverted on startup.
